### DISCUSSION TITLE
Database migration issues

### DISCUSSION THREAD

**Discussion by MaxiStefan** [simolus3/drift#2665] [Q&A] [answered]
Hi!

I had a few question about the onUpgrade callback that we can use to migrate the schema version. The current implemention is the following:

```

        onUpgrade: (Migrator m, int from, int to) async {
          await customStatement('PRAGMA foreign_keys = OFF');

          await transaction(
            () async {
              if (from < 2) {
                await migrateFromUnixTimestampsToText(m);
              }
              if (from < 3) {
                // project_devices table removed from version
                await m.deleteTable('project_devices');
              }
              if (from < 4) {
                final customPoleHeightColumn =
                    shapePoints.columnsByName['custom_pole_height'];
                if (customPoleHeightColumn == null) {
                  await m.addColumn(
                    shapePoints,
                    shapePoints.customPoleHeight,
                  );
                }
              }
              if (from < 5) {
                final appNameColumn = shapePoints.columnsByName['app_name'];
                final deviceNameColumn =
                    shapePoints.columnsByName['device_name'];
                final appVersionColumn =
                    shapePoints.columnsByName['app_version'];
                if (appNameColumn == null &&
                    deviceNameColumn == null &&
                    appVersionColumn == null) {
                  await m.addColumn(
                    shapePoints,
                    shapePoints.appName,
                  );
                  await m.addColumn(
                    shapePoints,
                    shapePoints.appVersion,
                  );
                  await m.addColumn(
                    shapePoints,
                    shapePoints.deviceName,
                  );
                }
              }
            },
          );
     }
```

I've seen in firebase situations where a crash would occur due to an attempt to **re-alter** the table to add the columns from then newly added ones in versions 4 and 5.(table columns were already where when the migration occured?How?)

To this end, i added the tableName.columnsByName['column_name'] and null check code so that I can work around the aforementioned problem. However, now I see insertion failures due to the columns not existing (migration has not occured? How)

Any ideas / suggestion as to what I might be doing wrong or how to fix the problems? 

Thanks in advance.

**Comment by simolus3** [comment, marked as answer]
> However, now I see insertion failures due to the columns not existing (migration has not occured? How)

`columnsByName` looks up columns as drift expects to be in the app's declared schema. So if `customPoleHeight` is accessible through the getter, `shapePoints.columnsByName['app_name']` will find it as well. `columnsByName` does not look up anything in the _database_ schema, so it doesn't indicate whether the column has actually been added.
In other words `== null` will always be false and the fixed migration won't add the column.

> (table columns were already where when the migration occured?How?)

This is a tougher one to answer. My best guess would be that the migration has finished in an incomplete state, but you've already wrapped everything in a transaction. I don't see how this could have happened from the snippet you've posted.

To fix the incorrect state, you would try wrapping `addColumn` in its own `try`/`catch` block. When it fails, it's probably due to the column already existing in the database. Alternatively, you could run a custom select for `select sql from sqlite_schema where name = 'shape_points'` and see if the `CREATE TABLE` statement that comes up contains the column name.

**Reply by MaxiStefan** [reply]
Right, I will be sure to test your suggestions. Thank you for the very quick response!
